packages/core/src/browser/request/browser-request-service.ts

```diff
@@ -0,0 +1,167 @@
+        return this.transformBackendResponse(backendResult);
+    }
+}
+
+@injectable()
+export class XHRBrowserRequestService extends DefaultBrowserRequestService {
+
+    protected authorization?: string;
+
+    override configure(config: RequestConfiguration): Promise<void> {
+        if ('proxyAuthorization' in config) {
+            this.authorization = config.proxyAuthorization;
+        }
+        return super.configure(config);
+    }
+
+    override async request(options: RequestOptions, token = CancellationToken.None): Promise<RequestContext> {
+        try {
+            const xhrResult = await this.xhrRequest(options, token);
+            const statusCode = xhrResult.res.statusCode ?? 200;
+            if (statusCode >= 400) {
+                return super.request(options);
+            }
+            return xhrResult;
+        } catch {
+            return super.request(options);
+        }
+    }
+
+    protected xhrRequest(options: RequestOptions, token: CancellationToken): Promise<RequestContext> {
+        const authorization = this.authorization || options.proxyAuthorization;
+        if (authorization) {
+            options.headers = {
+                ...(options.headers || {}),
+                'Proxy-Authorization': authorization
+            };
+        }
+
+        const xhr = new XMLHttpRequest();
+        return new Promise<RequestContext>((resolve, reject) => {
+
+            xhr.open(options.type || 'GET', options.url || '', true, options.user, options.password);
+            this.setRequestHeaders(xhr, options);
+
+            xhr.responseType = 'arraybuffer';
+            xhr.onerror = () => reject(new Error(xhr.statusText && ('XHR failed: ' + xhr.statusText) || 'XHR failed'));
+            xhr.onload = () => {
+                resolve({
+                    res: {
+                        statusCode: xhr.status,
+                        headers: this.getResponseHeaders(xhr)
+                    },
+                    buffer: new Uint8Array(xhr.response)
+                });
+            };
+            xhr.ontimeout = e => reject(new Error(`XHR timeout: ${options.timeout}ms`));
+
+            if (options.timeout) {
+                xhr.timeout = options.timeout;
+            }
+
+            xhr.send(options.data);
+
+            // cancel
+            token.onCancellationRequested(() => {
+                xhr.abort();
+                reject();
+            });
+        });
+    }
+
+    protected setRequestHeaders(xhr: XMLHttpRequest, options: RequestOptions): void {
+        if (options.headers) {
+            outer: for (const k of Object.keys(options.headers)) {
+                switch (k) {
+                    case 'User-Agent':
+                    case 'Accept-Encoding':
+                    case 'Content-Length':
+                        // unsafe headers
+                        continue outer;
+                }
+                xhr.setRequestHeader(k, options.headers[k]);
+            }
```

Comment:
Is the `outer` label necessary? Since `continue` doesn't apply to `switch` statements, I think that it will work without the label.